Sprint 3

Kosurij opened the pull request:
Добавлен роутинг, работа с апи, подробная валидация, уведомления при регистрации/добавлении пользователей/ошибках. 
Добавлена переписка по вебсокетам.

Commit:
Добавлен favicon.
Скорректирована валидация: выводятся понятные пользователю ошибки.
Добавлен класс роутера и изменен роутинг.

Commit:
Добавлена работа с HTTP и стейтом.

Commit:
Добавлена:
авторизацию в полном объеме (регистрация, авторизация, выход из системы);
работу с информацией пользователя (изменять данные пользователя, изменять пароль);

Commit:
Добавлено уведомление.

Commit:
Добавлена возможность изменять картинку

Commit:
Header вынесен в самостоятельный блок.

Commit:
Добавлено обновление аватара.

Commit:
Аватарка изменяется без перезагрузки страницы.

Commit:
Убрано сравнение isEqual

Commit:
Переделана структура чата

Commit:
Добавлена работа с чатами.

Commit:
Добавлен дропдаун

Commit:
Добавлена возможность добавления/удаления пользователя

Commit:
Подключены вебсокеты

Commit:
Прогнаны линтеры

Commit:
Обновлен ридми

Commit:
Файлу .stylelintrc добавлено расширение json

Commit:
Добавлены игноры

Commit:
Скорректирован жсдок

Commit:
Изменен package.json и netlify.toml

Commit:
Изменен package.json и netlify.toml

Commit:
Изменен netlify.toml

Commit:
Поднята версия ноды до 16.18.1

Commit:
Изменены стили

gennady-bars commented on src/utils/HTTPTransport.ts:
```diff
@@ -0,0 +1,86 @@
+}
+
+export default class HTTPTransport {
+  static API_URL = 'https://ya-praktikum.tech/api/v2';
```
Отлично, что базовый урл прописан 1 раз внутри класса

gennady-bars commented on src/utils/HTTPTransport.ts:
```diff
@@ -0,0 +1,86 @@
+    return this.request<Response>(this.endpoint + path);
+  };
+
+  public post<Response = void>(path: string, data?: TData): Promise<Response> {
```
**Можно лучше**
Если типизация повторяется в каждом методе, то нужно типизировать сам метод, а не дублировать типизацию аргументов. Код станет чище

```jsx

// создаем тип метода
type HTTPMethod<Response = void> = (url: string, options?: TData) => Promise<Response>

export class HTTPTransport {

  // используем тип и удаляем дублирование в аргументах
  get: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.GET}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  put: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.PUT}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  post: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.POST}, options.timeout)
  )
  // используем тип и удаляем дублирование в аргументах
  delete: HTTPMethod = (url, options = {}) => (
    this.request(url, {...options, method: METHODS.DELETE}, options.timeout)
  )

```

gennady-bars commented on src/utils/HTTPTransport.ts:
```diff
@@ -0,0 +1,86 @@
+  };
+
+  public get<Response>(path = ''): Promise<Response> {
+    return this.request<Response>(this.endpoint + path);
```
в методе `get` еще нужно обрабатывать `query`-параметры (если они переданы) с помощью функции `queryStringify` 


Вам ее давали в теории

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная



* `Readme` хорошо оформлен
* Проект задеплоен и отлично работает
* Отлично, что фиксируете версии зависимостей
* Отлично, что нет `EOF` ошибок в гите
* Хорошая структура папок и файлов
* Отличный роутинг
* Отличная логика сокетов
* Отлично, что базовый урл прописан 1 раз внутри класса

но есть некоторые недочеты: 


* при входе в чат в консоли ошибка https://disk.yandex.ru/i/4kRed2uLtzyV0w 
* Для любых запросов на сервер в **самом конце цепочки** запроса нужно осуществлять проверку на возможные ошибки с помощью `catch`. Можно `alert` добавить или `console.error`  **Нужно исправить это везде по коду в конце запросов.**
* в методе `get` еще нужно обрабатывать `query`-параметры (если они переданы) с помощью функции `queryStringify` 





**Можно лучше**

* нужна возможность изменить картинку чата. Есть такой функционал по задумке
* нужно иметь возможность удалить чат (можно добавить в меню сюда https://disk.yandex.ru/i/n1J9o2KC8VUVuQ )
* Если типизация повторяется в каждом методе, то нужно типизировать сам метод, а не дублировать типизацию аргументов. Код станет чище
* Попробуйте избавиться от `any`




Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте **работоспособность проекта** и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 

Напоминаю, что работа может быть принята только после исправления всех критических замечаний `Нужно исправить`.

Удачного рефакторинга кода.

Commit:
Убрал мета заголовок, вызывающий ошибку в консоли.
Добавил обработку гет параметров.
Обернул все запросы к серверу в контроллерах в функцию makeRequest, которая содержит в себе try/catch.

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.

Kosurij commented:
Merge sprint_3 PR
